packages/ozone/src/set/service.ts

```diff
@@ -0,0 +1,227 @@
+      (eb) =>
+        eb
+          .selectFrom('set_value')
+          .select((e) => e.fn.count<number>('setId').as('count'))
+          .whereRef('setId', '=', 's.id')
+          .as('setSize'),
+    ])
+  }
+
+  async query({
+    limit,
+    cursor,
+    namePrefix,
+    sortBy,
+    sortDirection,
+  }: {
+    limit: number
+    cursor?: string
+    namePrefix?: string
+    sortBy: 'name' | 'createdAt' | 'updatedAt'
+    sortDirection: 'asc' | 'desc'
+  }): Promise<{
+    sets: Selectable<SetDetail & { setSize: number }>[]
+    cursor?: string
+  }> {
+    let qb = this.buildQueryForSetWithSize().limit(limit)
+
+    if (namePrefix) {
+      qb = qb.where('s.name', 'like', `${namePrefix}%`)
+    }
+
+    if (cursor) {
+      if (sortBy === 'name') {
+        qb = qb.where('s.name', sortDirection === 'asc' ? '>' : '<', cursor)
+      } else {
+        qb = qb.where(
+          `s.${sortBy}`,
+          sortDirection === 'asc' ? '>' : '<',
+          new Date(cursor),
+        )
+      }
+    }
+
+    qb = qb.orderBy(`s.${sortBy}`, sortDirection)
+
+    const sets = await qb.execute()
+    const lastItem = sets.at(-1)
+
+    return {
+      sets,
+      cursor: lastItem
+        ? sortBy === 'name'
+          ? lastItem?.name
+          : lastItem?.[sortBy].toISOString()
+        : undefined,
+    }
+  }
+
+  async getByName(name: string): Promise<Selectable<SetDetail> | undefined> {
+    const query = this.db.db
+      .selectFrom('set_detail')
+      .selectAll()
+      .where('name', '=', name)
+
+    return await query.executeTakeFirst()
+  }
+
+  async getByNameWithSize(
+    name: string,
+  ): Promise<Selectable<SetDetail & { setSize: number }> | undefined> {
+    return await this.buildQueryForSetWithSize()
+      .where('s.name', '=', name)
+      .executeTakeFirst()
+  }
+
+  async getSetWithValues({
+    name,
+    limit,
+    cursor,
+  }: {
+    name: string
+    limit: number
+    cursor?: string
+  }): Promise<
+    | {
+        set: Selectable<SetDetail & { setSize: number }>
+        values: string[]
+        cursor?: string
+      }
+    | undefined
+  > {
+    const set = await this.getByNameWithSize(name)
+    if (!set) return undefined
+
+    const { ref } = this.db.db.dynamic
+    let qb = this.db.db
+      .selectFrom('set_value')
+      .selectAll()
+      .where('setId', '=', set.id)
+
+    const keyset = new TimeIdKeyset(ref(`createdAt`), ref('id'))
+    const paginatedBuilder = paginate(qb, {
+      limit,
+      cursor,
+      keyset,
+      direction: 'asc',
+    })
+
+    const result = await paginatedBuilder.execute()
+
+    return {
+      set,
+      values: result.map((v) => v.value),
+      cursor: keyset.packFromResult(result),
+    }
+  }
+  async upsert({
+    name,
+    description,
+  }: Pick<SetDetail, 'name' | 'description'>): Promise<void> {
+    await this.db.db
+      .insertInto('set_detail')
+      .values({
+        name,
+        description,
+        updatedAt: new Date(),
```

Comment:
I think you have postgres setup to set this field automatically on insert (same as `createdAt`).